Add named frames to CollisionObjects

felixvd opened the pull request:
### Description

This PR implements #1041. It adds named frames to collision objects so that they can be used for planning, both as targets in the planning scene, as well as end effector links when the objects are attached to the robot. This allows writing commands like "Move the tip of the screwdriver to the head of the screw" or "Move the spout of the teapot to the top of the mug". It also opens up avenues for using in-hand pose estimation to update the pose of an object in the gripper (e.g. grasp a tool, confirm the object pose with a camera, then use the updated tool tip pose on the object).

Currently, the named frames are displayed in Rviz in the Scene Objects tab and as red dots on the `AttachedBodies`. I also made a header file that converts URDF files to `collision_object` messages in C++, but it's not part of this PR. I didn't test it a lot yet, so there are surely some bugs, but it should be good enough to try out.

[Here](https://gist.github.com/felixvd/9b79be2bd10cc615e980f389bab2096d) is a gist that you can use to create the following example images. These run for a UR5, but you have to adjust everything to your system and fill in some variables. I didn't have time to prepare a whole test package.

![2018-9-11-ps-extension](https://user-images.githubusercontent.com/4535737/45336934-a9af8200-b5c0-11e8-85be-31d180b2a07c.png)

The result of "Move the cylinder tip to the bottom of the box":

![2018-09-11-ps-extension-2](https://user-images.githubusercontent.com/4535737/45336938-ac11dc00-b5c0-11e8-8e5d-71c56015c264.png)

Open ToDos:

* Test the path and trajectory constraints (I only checked that it works on goal constraints)
* Integrate the frames' visualization into Rviz (should they be published to TF? Should there be a checkbox?)
* Implement MOVE for CollisionObjects and AttachedBodies (would be great for updating the object pose in the gripper)
* Check effect on performance

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Include a screenshot if changing a GUI
- [ ] Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Decide if this should be cherry-picked to other current ROS branches
- [ ] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

v4hn commented:
Wow, this is quite a lot of work :1st_place_medal: 
This takes some time to review & refine and I will sadly not be able to do this in the next month.

We shortly talked about this before and, unless I'm mistaken now,
we also talked about the alternative to add relative transforms to collision-objects, essentially allowing to add empty collision objects for relative frames, too, in a style similar to scene graphs.
Could you say a few words about why you chose child links instead?

Afaics now this will leave us with two somewhat redundant mechanisms if we should ever get around to implement scene graph support (which has been requested before).

v4hn commented:
Oh, one more important point.
This change requires message changes, so sadly we cannot merge it into the `kinetic-devel` branch anymore!

Personally I believe `melodic-devel` is still quite fine, but others might disagree.

felixvd commented:
I agree that a proper scene graph would be good. The main reason I went this way is that I wasn't sure how much effort it would be, and this seemed like a less intrusive change. It also solves most of my application's pain points with a conceptually simple change, which was an argument in favor.

With the child objects/scene graph structure we discussed, I worried about additional pitfalls (recursive transforms and message processing). I think I might be able to tackle it now that I know a bit more, but I didn't want to bite off more than I could chew. And I am still not sure if I really know how the world object and AttachedBody data are communicated to different parts of the system.

While the child link structure may be redundant in some ways in the future, it is readable: One physical object is contained in a single CollisionObject message, and the intention of a named frame is clear. If scene graph support is added in the way you envision and all child frames are collision objects, an object with child frames would be spread over multiple CollisionObject messages, which seems less convenient and harder to read. That said, I'm totally open to doing and arranging things differently.

Lastly, this change doesn't get in the way of future work, and could coexist with a scene graph, as you said. The meat of commit `defbd63` does not depend on the child links either and its extension to a scene-graph-like structure is straight forward.

If this can't be merged into `kinetic-devel`, let me know how to proceed. I suppose the missing messages are the reason the checks are failing.

davetcoleman commented:
Really great work! More thoughts, beyond those in the related issue https://github.com/ros-planning/moveit/issues/1041 you created:

> Here is a gist that you can use to create the following example images.

This needs to be updated in the tutorial, either a new one or add to an existing one

>  I didn't have time to prepare a whole test package.

This has some big changes to MoveIt!, adding a few tests would make it much easier to get in

davetcoleman commented on moveit_core/robot_state/include/moveit/robot_state/attached_body.h:
```diff
@@ -1,36 +1,36 @@
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
```
this kind of change should be avoided

felixvd replied:
Pretty sure this is done with the clang-format now. Just checking these off so I can keep track of open comments.

davetcoleman commented on moveit_ros/move_group/src/default_capabilities/constraint_validation_service_capability.cpp:
```diff
@@ -0,0 +1,77 @@
+  // for (std::size_t i = 0; i < res.constraints.size(); ++i)
+  // {
+    // v = v * kinematic_constraints::validatePositionOrientationConstraints(rs, res.constraints[i]);
+  // }
```
cleanup?

davetcoleman commented on moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp:
```diff
@@ -288,10 +291,11 @@ class MoveGroupInterface::MoveGroupInterfaceImpl
+                                               "replaced by an action)."
+                                               "\nReplace 'MoveGroupExecuteService' with "
+                                               "'MoveGroupExecuteTrajectoryAction' in "
+                                               "move_group.launch");
```
are you using a different version of clang-format? these changes should not happen

davetcoleman commented on moveit_ros/visualization/rviz_plugin_render_tools/src/robot_state_visualization.cpp:
```diff
@@ -132,6 +132,18 @@ void RobotStateVisualization::updateHelper(const robot_state::RobotStateConstPtr
     }
+
+    // Insert the named frames as red spheres of radius 2 mm
+    // TODO: Integrate this properly (either by publishing to TF, or frames that can be toggled)
```
add your name to the TODO 

TODO(username)

davetcoleman commented on moveit_ros/visualization/rviz_plugin_render_tools/src/robot_state_visualization.cpp:
```diff
@@ -132,6 +132,18 @@ void RobotStateVisualization::updateHelper(const robot_state::RobotStateConstPtr
+
+    // Insert the named frames as red spheres of radius 2 mm
+    // TODO: Integrate this properly (either by publishing to TF, or frames that can be toggled)
+    auto s = shapes::Sphere(.002);
```
longer var names

davetcoleman commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,51 @@ bool World::hasObject(const std::string& id) const
+  // Check object names first
+  if (objects_.find(id) != objects_.end())
+    return true;
+  else  // Then objects' named frames
```
no need for else here

felixvd replied:
True, but I tend to use explicit else blocks because I find the intent clearer. Is there a style guideline where this is decided?

Some more words on this: https://softwareengineering.stackexchange.com/questions/255523/do-else-blocks-increase-code-complexity

felixvd replied:
Any input from others on this?

davetcoleman commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,51 @@ bool World::hasObject(const std::string& id) const
+  auto it = objects_.find(id);
+  if (it != objects_.end())
+    return it->second->shape_poses_[0];
+  else  // Find within named frames
```
remove else

felixvd replied:
See comment above

davetcoleman commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,51 @@ bool World::hasObject(const std::string& id) const
+
+std::string World::getParent(const std::string& id) const
+{
+  for (auto o : objects_)
```
longer var name than 'o'

davetcoleman commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -273,3 +273,156 @@ moveit_msgs::Constraints kinematic_constraints::constructGoalConstraints(const s
+bool kinematic_constraints::validatePositionConstraints(const robot_state::RobotState& state,
+                                                        moveit_msgs::Constraints& c)
+{
+  ROS_INFO("Validating position constraints.");
```
add NAMED to these INFOs

felixvd replied:
My bad, this file is the only one where I left my actual debugging lines. Will update after ROSCon JP.

felixvd replied:
When no one is looking, I even use "print()" because it's quicker to type than "rospy.loginfo()".

davetcoleman commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -273,3 +273,156 @@ moveit_msgs::Constraints kinematic_constraints::constructGoalConstraints(const s
+      }
+
+      if (transform_found)
+      {
```
Remove the else and just use:

      if (!transform_found)
           return false

davetcoleman commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -83,14 +83,9 @@ class SceneTransforms : public robot_state::Transforms
 
 private:
-  bool knowsObject(const std::string& id) const
+  bool knowsFrame(const std::string& id) const
```
add inline comments about why its Frame now and not Object

felixvd replied:
It's because this function seemed to be used to check if the transform to a frame is known, rather than to check if the object exists in the world, but I think there is a use case for both. Either way, before changing the name the original function should be deprecated. I readded both.

felixvd replied:
Correction: I changed knowsObject slightly so it returns true when the object is known to the world (named frames are ignored). The previous version of the function makes more sense as "knowsFrame". I am not sure what I think of only checking the world for an object, and ignoring the attachedBody objects, but that's another discussion.

davetcoleman commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -837,6 +835,14 @@ bool PlanningScene::getCollisionObjectMsg(moveit_msgs::CollisionObject& collisio
+  for (auto it : obj->named_frames_)
+  {
+    collision_obj.frame_names.push_back(it.first);
+    geometry_msgs::Pose p;
```
longer var names

davetcoleman commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1423,9 +1430,16 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
   if (object.object.operation == moveit_msgs::CollisionObject::ADD ||
       object.object.operation == moveit_msgs::CollisionObject::APPEND)
   {
+    ////----------------------
```
remove messy comment?

davetcoleman commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1447,65 +1461,58 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
 
-      // we need to add some shapes; if the message is empty, maybe the object is already in the world
+      // STEP 1: Get info about object from the world. First shapes, then named frames.
+      // TODO: This code may be duplicated in robot_state/conversions.cpp
```
in a separate PR can you unify these?

davetcoleman commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1790,6 +1871,7 @@ bool PlanningScene::processCollisionObjectMsg(const moveit_msgs::CollisionObject
                         new_poses.size(), object.id.c_str(), obj->shapes_.size());
         return false;
       }
+      // TODO: Move the named frames at least!!
```
??

davetcoleman commented on moveit_core/robot_state/src/conversions.cpp:
```diff
@@ -99,8 +99,9 @@ static bool _multiDOFJointsToRobotState(const sensor_msgs::MultiDOFJointState& m
-                                    "but it was not possible to transform that to frame '%s'",
+      ROS_WARN_NAMED("robot_state",
+                     "The transform for multi-dof joints was specified in frame '%s' "
+                     "but it was not possible to transform that to frame '%s'",
```
why did all these formatting changes happen?

davetcoleman commented on moveit_core/robot_state/src/conversions.cpp:
```diff
@@ -215,6 +216,15 @@ static void _attachedBodyToMsg(const AttachedBody& attached_body, moveit_msgs::A
       sv.addToObject(sm, p);
     }
   }
+  aco.object.frame_names.clear();
```
add comments for what this new code does

felixvd replied:
Does it really belong there? This just implements the message. I think comments explaining it should go in the message definition.

davetcoleman commented on moveit_core/robot_state/src/robot_state.cpp:
```diff
@@ -128,9 +128,10 @@ void RobotState::copyFrom(const RobotState& other)
+    memcpy(position_, other.position_,
+           robot_model_->getVariableCount() * sizeof(double) *
+               (1 + ((has_velocity_ || has_acceleration_ || has_effort_) ? 1 : 0) +
+                ((has_acceleration_ || has_effort_) ? 1 : 0)));
```
these changes should not be here

mlautman replied:
@felixvd Are you sure you are using the correct clang formatting?

felixvd replied:
No, I ran it from inside Visual Studio Code. It probably wasn't the right version.

davetcoleman commented on moveit_ros/move_group/src/default_capabilities/constraint_validation_service_capability.cpp:
```diff
@@ -0,0 +1,77 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2012, Willow Garage, Inc.
```
update

davetcoleman commented on moveit_ros/move_group/src/default_capabilities/constraint_validation_service_capability.cpp:
```diff
@@ -0,0 +1,77 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Willow Garage nor the names of its
```
update

davetcoleman commented on moveit_ros/move_group/src/default_capabilities/constraint_validation_service_capability.cpp:
```diff
@@ -0,0 +1,77 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Felix von Drigalski */
```
add file description please

davetcoleman commented on moveit_ros/move_group/src/move_group.cpp:
```diff
@@ -112,8 +113,12 @@ class MoveGroupExe
+                                                                                                               "MoveGro"
+                                                                                                               "upCapab"
+                                                                                                               "ilit"
+                                                                                                               "y"));
```
restore

davetcoleman commented on moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame_objects.cpp:
```diff
@@ -160,7 +160,7 @@ static QString decideStatusText(const collision_detection::CollisionWorld::Objec
 {
```
i think changes to Rviz should be a follow-up PR to keep the size of this one manageable. feel free to create it now, though

felixvd replied:
Sure? It would make testing this PR less convenient, and I hope the Rviz changes will stay pretty limited.

davetcoleman replied:
its not a hard rule, just a would be nice. your argument is fine

felixvd replied:
#1122 and the ToDo in the code make a good case for rethinking the Rviz changes and how to visualize these frames. Red dots are obviously not ideal. I will keep this in for testing and remove it before the merge.

davetcoleman commented:
The recommended command line command for using the correct clang format is here:
http://moveit.ros.org/documentation/contributing/code/

felixvd commented:
Thanks, that helps. I'm in a crunchtime I had to skip ROSCon and IROS for, but I'll get back to this at the latest after 20 October and before MoveIt Day.

As a quick note to myself, this is not compatible with planCartesianPath yet.

mlautman commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1728,6 +1799,17 @@ bool PlanningScene::processCollisionObjectMsg(const moveit_msgs::CollisionObject
+    {
+      tf::poseMsgToEigen(object.named_frames[i], p);
+      std::string name = object.frame_names[i];
+      named_frames[name] = t * p;
```
This is interesting, can you explain why you chose to transform the named_frames into the world frame rather than leaving them wrt their parent frame? It seems like it may be less work to keep the transforms as "the tip is some pose from the object" rather than have to update the tip transform every time you move the object in the planning scene.

felixvd replied:
It was so that getFrameTransform() and knowsFrameTransform() would apply the same way as they do to the other transforms in the world. I just copy-pasted it from the lines above, because I wanted to change as little as possible. I agree, it looks inefficient, but the same argument may apply to the other poses (primitives/meshes/planes).

But it is also true that the other poses are relevant for collision checking, and these frames are not, so keeping the named frames in the object reference frame might be better. I don't know if that makes a difference though - I haven't read the collision checking code.

mlautman commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1728,6 +1799,17 @@ bool PlanningScene::processCollisionObjectMsg(const moveit_msgs::CollisionObject
+    // Add named frames
+    std::map<std::string, Eigen::Affine3d> named_frames;
+    Eigen::Affine3d p;
+    for (std::size_t i = 0; i < object.named_frames.size(); ++i)
```
Unfortunately, we can't update the messages in kinetic-devel without breaking the release. Can we target melodic with these changes.

felixvd replied:
I am not sure anymore what the conclusion of our discussion regarding this was @davetcoleman. I don't have an issue with targeting only melodic.

mlautman commented on moveit_core/robot_state/src/conversions.cpp:
```diff
@@ -215,6 +216,15 @@ static void _attachedBodyToMsg(const AttachedBody& attached_body, moveit_msgs::A
+  for (auto it : attached_body.getNamedTransforms())
+  {
+    aco.object.frame_names.push_back(it.first);
+    geometry_msgs::Pose p;
```
longer variable name

mlautman commented on moveit_core/robot_state/src/conversions.cpp:
```diff
@@ -215,6 +216,15 @@ static void _attachedBodyToMsg(const AttachedBody& attached_body, moveit_msgs::A
   }
+  aco.object.frame_names.clear();
+  aco.object.named_frames.clear();
+  for (auto it : attached_body.getNamedTransforms())
```
better variable name

mlautman commented on moveit_ros/move_group/src/default_capabilities/constraint_validation_service_capability.cpp:
```diff
@@ -0,0 +1,77 @@
+  // Checks if the goal is defined either for existing link_names or the names of AttachedBody objects
+  // that are attached to the robot. The latter are transformed to link_names so the planner can deal with them.
+  res.constraints = req.constraints;
+  bool v = true;
```
valid

mlautman commented on moveit_ros/move_group/src/default_capabilities/constraint_validation_service_capability.cpp:
```diff
@@ -0,0 +1,77 @@
+  planning_scene_monitor::LockedPlanningSceneRO ls(context_->planning_scene_monitor_);
+  robot_state::RobotState rs = ls->getCurrentState();
+
+  ROS_INFO("Validating constraints inside capability");
```
NAMED

mlautman commented on moveit_ros/move_group/src/default_capabilities/constraint_validation_service_capability.cpp:
```diff
@@ -0,0 +1,77 @@
+  // {
+    // v = v * kinematic_constraints::validatePositionOrientationConstraints(rs, res.constraints[i]);
+  // }
+  v = v * kinematic_constraints::validatePositionOrientationConstraints(rs, res.constraints);
```
`v = v && ...`?

felixvd replied:
That's way nicer, thanks

felixvd commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,51 @@ bool World::hasObject(const std::string& id) const
+  {
+    for (auto o : objects_)
+    {
+      if (o.second->named_frames_.count(id) > 0)
```
It's true that every frame name has to be unique, just like TF frames. I will make that clear in the message description.

I don't think this line here is prone to bugs apart from that, though.

rhaschke replied:
I think, Mike's comment emphasizes an important implementation issue: While the implementation, using an per-object-map, suggests that we can use the same frame name multiple times (for different objects), actually we can't. It would be better, if this fact would be explicitly reflected in the implementation as well, e.g. using a single global map from frame names to (object, pose) pairs.

felixvd replied:
Hmm, I wouldn't like to rewrite that. It would indeed be much nicer to be able to reuse the same frame name on different objects, but we would have to define how the object namespaces its frames. How is it done in tf2? I think it would be best to stay as close to that convention as possible, so that the frames can (in a final step) be published to tf2 as well.

felixvd replied:
Just for reference, the description I just added to the message:

```
# The names of named frames in the scene need to be unique! If both "screwdriver" and "pencil" have
# a named framed called "tip", it will cause silent errors! Use "screwdriver_tip" and "pencil_tip" instead,
# or submit a pull request to fix this limitation.
```

felixvd commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -169,6 +214,18 @@ bool World::moveObject(const std::string& id, const Eigen::Affine3d& transform)
   return true;
 }
 
+bool World::replaceShapesInObject(const std::string& id, const std::vector<shapes::ShapeConstPtr>& shapes,
```
The original functions refer to the objects with "id". I would stay consistent, whatever the choice.

felixvd replied:
I changed all of them to object_id

felixvd commented on moveit_core/kinematic_constraints/include/moveit/kinematic_constraints/utils.h:
```diff
@@ -183,6 +183,39 @@ moveit_msgs::Constraints constructGoalConstraints(const std::string& link_name,
-}
+
+/**
+ * \brief Attempts to make a constraint message valid which in its link_name field
```
How about this?
"Converts constraints that refer to frames of AttachedBody objects to frames that exist on the robot. The link_name field is changed to the link that the AttachedBody object is attached to."

felixvd replied:
Changed to:
"Converts constraints that refer to frames of AttachedBody objects to frames that exist on the robot. This is required to construct a valid planning request. The link_name field of the constraint is changed from the name of the object frame to the name of the robot link that the  AttachedBody object is attached to.
This function is used when constructing goals. "

felixvd commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -273,3 +273,156 @@ moveit_msgs::Constraints kinematic_constraints::constructGoalConstraints(const s
+      if (state.hasAttachedBody(ori_con_it->link_name))
+      {
+        ROS_INFO_STREAM("State has attached body with name " << ori_con_it->link_name);
+        auto ab = state.getAttachedBody(ori_con_it->link_name);
```
I just don't like reading the same verbose definitions every time. I don't have a strong opinion on this though, so I can be more explicit if that's the consensus.

felixvd commented:
As discussed with Dave on Wednesday, I had to freeze this until the week of the MoveIt Day because I am completely occupied with [this](http://worldrobotsummit.org/en/programs/challenge/). I think I can get it done and tested on Melodic by WMD though.

rhaschke commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,51 @@ bool World::hasObject(const std::string& id) const
+    return true;
+  else  // Then objects' named frames
+  {
+    for (auto o : objects_)
```
I think, for an iterator variable, o is fine here.

felixvd replied:
I unified this naming to the rest of the file.

rhaschke commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,51 @@ bool World::hasObject(const std::string& id) const
+      return o.second->id_;
+  }
+  // If it is an object's name, return the object
+  auto it = objects_.find(id);
```
Here, you change the order (now first searching for object frames, then for objects).
In all other functions, you did it the other way around. This can introduce bugs with non-unique names as well.

felixvd replied:
True. It's because this is the only one I wrote from scratch, as a simple getter function. I am not sure it is necessary at all. I changed the order, but I would also be ok with deleting the function.

felixvd replied:
Reordered this as requested.

rhaschke commented on moveit_core/collision_detection/include/moveit/collision_detection/world.h:
```diff
@@ -160,6 +176,13 @@ class World
   bool moveObject(const std::string& id, const Eigen::Affine3d& transform);
 
+  /** \brief Replaces all shapes in an existing object.
+   * If the object already exists, this call will add the shape to the object
```
Looking at the implementation, I found that the documentation here is wrong, but the function name matches implementation.

felixvd replied:
True. I changed the documentation.

rhaschke commented on moveit_core/robot_state/include/moveit/robot_state/attached_body.h:
```diff
@@ -110,6 +111,47 @@ class AttachedBody
   }
 
+  /** \brief Get the fixed transform to a named frame on this body (not a transform to a visual or collision shape) */
+  const Eigen::Affine3d& getFixedTransform(const std::string& frame_name) const
```
Why introducing an alias for getNamedTransform?

felixvd replied:
I probably imagined using it like `getFrameTransform` in `robot_state.cpp` to look in both the named frames and the shapes' frames, but I don't think it will be needed. 

It probably came to my mind because `getFixedTransforms` is a bit of an awkward name. Maybe `getGeometryTransforms` or `getShapeTransforms` might be better?

felixvd replied:
I couldn't follow my reasoning and removed the alias. It was unused.

rhaschke commented on moveit_core/robot_state/include/moveit/robot_state/attached_body.h:
```diff
@@ -110,6 +111,47 @@ class AttachedBody
+   */
+  bool hasNamedTransform(const std::string& frame_name) const
+  {
+    return (named_frames_.count(frame_name) > 0);
```
Better use `find(frame_name) != end()`. `count()` needs to process all elements.

felixvd replied:
Thanks, very helpful

felixvd commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,51 @@ bool World::hasObject(const std::string& id) const
+    return true;
+  else  // Then objects' named frames
+  {
+    for (auto o : objects_)
```
I agreed with what you wrote, but the rest of the file seems to follow the same convention. I only changed from `auto` to `auto const&` for now, under the assumption that it's faster.

felixvd commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -401,8 +396,9 @@ PlanningScene::getCollisionRobotUnpadded(const std::string& collision_detector_n
   {
-    ROS_ERROR_NAMED("planning_scene", "Could not get CollisionRobotUnpadded named '%s'. "
-                                      "Returning active CollisionRobotUnpadded '%s' instead",
+    ROS_ERROR_NAMED("planning_scene",
```
Agreed. Maybe @mlautman can post an issue and then resolve this comment

felixvd commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -837,6 +835,14 @@ bool PlanningScene::getCollisionObjectMsg(moveit_msgs::CollisionObject& collisio
     if (hasObjectType(collision_obj.id))
       collision_obj.type = getObjectType(collision_obj.id);
   }
+  for (auto it : obj->named_frames_)
```
Changed to frame_pair

felixvd commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1447,65 +1461,58 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
+      return false;
+    }
+
     const robot_model::LinkModel* lm = getRobotModel()->getLinkModel(object.link_name);
```
Done

felixvd commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1447,65 +1461,58 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
 
-          // need to transform poses to the link frame
+          // Transform shape poses to the link frame
           const Eigen::Affine3d& i_t = kstate_->getGlobalLinkTransform(lm).inverse();
```
Changed to inv_transform to align with a change further down

felixvd commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1551,42 +1558,98 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
       ROS_ERROR_NAMED("planning_scene", "Robot state is not compatible with robot model. This could be fatal.");
   }
-  else if (object.object.operation == moveit_msgs::CollisionObject::REMOVE)
+  else if (object.object.operation == moveit_msgs::CollisionObject::REMOVE)  // == DETACH
```
I added it as an explanation because for attached objects this really means "detach" instead of "remove from the scene". I would be in favor of making "DETACH" an explicit codeword, but that's for a different PR.

felixvd commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1728,6 +1799,17 @@ bool PlanningScene::processCollisionObjectMsg(const moveit_msgs::CollisionObject
+    {
+      tf::poseMsgToEigen(object.named_frames[i], p);
+      std::string name = object.frame_names[i];
+      named_frames[name] = t * p;
```
changed to the very verbose "transform_to_obj_reference_frame * pose" (pose refers to to the pose of the shape primitive/mesh/plane in the reference frame)

felixvd commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1778,9 +1860,8 @@ bool PlanningScene::processCollisionObjectMsg(const moveit_msgs::CollisionObject
       if (obj->shapes_.size() == new_poses.size())
       {
         std::vector<shapes::ShapeConstPtr> shapes = obj->shapes_;
+        world_->replaceShapesInObject(object.id, shapes, new_poses);
```
I am looking at this previous code, where `removeObject` also returned `false` if the object.id isn't found: https://github.com/ros-planning/moveit/blob/melodic-devel/moveit_core/planning_scene/src/planning_scene.cpp#L1814

I don't think this changes anything in that regard. In either case, the return value gets ignored in either version. Maybe I am misunderstanding something?

If you mean the `processCollisionObjectMsg` function, I don't think the return value of that has changed.

felixvd commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -215,6 +272,17 @@ void World::clearObjects()
 
+bool World::setNamedFramesOfObject(const std::string& id, const std::map<std::string, Eigen::Affine3d>& named_frames)
+{
+  ObjectPtr& obj = objects_[id];
```
Very true. Corrected.

felixvd commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -215,6 +272,17 @@ void World::clearObjects()
   objects_.clear();
 }
 
+bool World::setNamedFramesOfObject(const std::string& id, const std::map<std::string, Eigen::Affine3d>& named_frames)
```
I can imagine some use cases, but I don't think it's required for a first version. Do you have anything specific in mind?

felixvd commented:
A quick update before I get on my plane. I have been on the road constantly and work from my laptop's VM is slow, but I added most of the comments during WMD. The conversations related to those are marked as resolved. The last commit changes a lot of lines, but (hopefully) very little functionally.

There are some decisions to take that affect the work left to do on this PR:  
1) Does the constraint validation have to be in a planning request adapter, or do we accept it as is for now? I am a bit worried that moving the validation to an adapter might take a while because I have never touched that part of the pipeline, but maybe it is easier than I think. At the same time, this could be changed later rather than now.   
2) Should the list of named frames be global for the scene, or held by each object? I would be in favor of the latter, because it is already implemented and it will allow multiple objects with same-name frames after we decide how to namespace them. 

Known limitations:
- The current version of planCartesianPath is not supported because it does not work the same way as the other planning requests.  
- There is no visualization of the frames in Rviz because they are not published to TF (the temporary visualization I added should be taken out before the merge)  

Open small ToDos:
- A `getFrameInfo()` method as suggested by Robert
- Combining the constraint validation methods
- Rebase the branch and resolve the conflicts

Another note: I will be in Europe for 3 weeks. That's another reason I am reluctant to introduce bigger changes now. I hope I can finish the small stuff very shortly, though. Will update after my flight!

felixvd commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -273,3 +273,156 @@ moveit_msgs::Constraints kinematic_constraints::constructGoalConstraints(const s
+  return true;
+}
+
+bool kinematic_constraints::validateOrientationConstraints(const robot_state::RobotState& state,
```
As mentioned above, it went into `getFrameInfo()` as you suggested.

felixvd commented:
Closing because it is superceded by #1439.
